poncho/src/poncho/library_network_code.py

```diff
@@ -1,414 +1,460 @@
-
-            # in case of FutureFunctionCall tasks
-            new_args = []
-            for arg in args:
-                if isinstance(arg, dict) and "VineFutureFile" in arg:
-                    with open(arg["VineFutureFile"], "rb") as f:
-                        output = cloudpickle.load(f)["Result"]
-                        new_args.append(output)
-                else:
-                    new_args.append(arg)
-            args = tuple(new_args)
 
-            try:
-                result = func(*args, **kwargs)
-                success = True
-                reason = None
-            except Exception:
-                result = None
-                success = False
-                reason = traceback.format_exc()
-            return LibraryResponse(result, success, reason).generate()
-
-        return remote_wrapper
-
-    # Handler to sigchld when child exits.
-    def sigchld_handler(signum, frame):
-        # write any byte to signal that there's at least 1 child
-        os.writev(w, [b"a"])
-
-    # Read data from worker, start function, and dump result to `outfile`.
-    def start_function(in_pipe_fd, thread_limit=1):
-        # read length of buffer to read
-        buffer_len = b""
-        while True:
-            c = os.read(in_pipe_fd, 1)
-            if c == b"":
-                stdout_timed_message(f"can't get length from in_pipe_fd {in_pipe_fd}")
-                exit(1)
-            elif c == b"\n":
-                break
+# import relevant libraries.
+import json
+import os
+import fcntl
+import sys
+import argparse
+import traceback
+import cloudpickle
+import select
+import signal
+import time
+from datetime import datetime
+import socket
+from threadpoolctl import threadpool_limits
+from ndcctools.taskvine.utils import load_variable_from_library
+
+# self-pipe to turn a sigchld signal when a child finishes execution
+# into an I/O event.
+r, w = os.pipe()
+exec_method = None
+
+
+# This class captures how results from FunctionCalls are conveyed from
+# the library to the manager.
+# For now, all communication details should use this class to generate responses.
+# In the future, this common protocol should be listed someplace else
+# so library tasks from other languages can use.
+class LibraryResponse:
+    def __init__(self, result=None, success=None, reason=None):
+        self.result = result
+        self.success = success
+        self.reason = reason
+
+    def generate(self):
+        return {
+            "Result": self.result,
+            "Success": self.success,
+            "Reason": self.reason,
+        }
+
+
+# A wrapper around functions in library to extract arguments and formulate responses.
+def remote_execute(func):
+    def remote_wrapper(event):
+        args = event.get("fn_args", [])
+        kwargs = event.get("fn_kwargs", {})
+
+        # in case of FutureFunctionCall tasks
+        new_args = []
+        for arg in args:
+            if isinstance(arg, dict) and "VineFutureFile" in arg:
```

Comment:
This smells like trouble to me.  Why does the worker need to know if this is a future file as opposed to the creation of a regular file?

Reply:
I think this belongs to the future function calls and I don't modify that code except left-indent that code block in. @JinZhou5042 I see that you added this line in https://github.com/cooperative-computing-lab/cctools/pull/3670. Can you shed some insights please?